handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java

```diff
@@ -0,0 +1,163 @@
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.ssl.util;
+
+import io.netty.util.CharsetUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.security.AccessController;
+import java.security.PrivateKey;
+import java.security.PrivilegedAction;
+import java.security.Provider;
+
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.openssl.PEMDecryptorProvider;
+import org.bouncycastle.openssl.PEMEncryptedKeyPair;
+import org.bouncycastle.openssl.PEMException;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.bouncycastle.openssl.jcajce.JceOpenSSLPKCS8DecryptorProviderBuilder;
+import org.bouncycastle.openssl.jcajce.JcePEMDecryptorProviderBuilder;
+import org.bouncycastle.operator.InputDecryptorProvider;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.pkcs.PKCS8EncryptedPrivateKeyInfo;
+import org.bouncycastle.pkcs.PKCSException;
+
+public final class BouncyCastlePemReader {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(BouncyCastlePemReader.class);
+    private static Provider bcProvider;
+    private static Boolean available;
+
+    static {
+        tryLoading();
+    }
+
+    public static boolean isAvailable() {
+        return available;
+    }
+
+    private static void tryLoading() {
+        available = AccessController.doPrivileged(new PrivilegedAction<Boolean>() {
+            @Override
+            public Boolean run() {
+                try {
+                    Class<Provider> bcProviderClass
+                      = (Class<Provider>) Class.forName("org.bouncycastle.jce.provider.BouncyCastleProvider",
+                              true, this.getClass().getClassLoader());
+                    bcProvider = bcProviderClass.newInstance();
+                    logger.debug("Bouncy Castle provider available");
+                    return Boolean.TRUE;
+                } catch (Exception e) {
+                    logger.debug("Cannot load Bouncy Castle provider", e);
```

Comment:
Exception / Errors should be saved, likely in some sort of bouncyCastleUnavailabilityCause field for later use.

Reply:
+1. Look at the native transports for an example of this (e.g. [KQueue](https://github.com/netty/netty/blob/4.1/transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueue.java#L27))